projects/storefrontlib/src/shared/components/popover/popover.directive.ts

```diff
@@ -0,0 +1,207 @@
+   */
+  @Output() closePopover?: EventEmitter<void> = new EventEmitter();
+
+  /**
+   * Flag used to inform about current state of popover component.
+   * Popover is closed by default, so value is set to false.
+   */
+  isOpen: boolean;
+
+  /**
+   * Popover component instance.
+   */
+  popoverContainer: ComponentRef<PopoverComponent>;
+
+  /**
+   * Configuration for a11y improvements.
+   */
+  focusConfig: FocusConfig;
+
+  /**
+   * Listen events fired on element binded to popover directive.
+   *
+   * Based on event type some a11y improvements can be made.
+   * For example if popover was opened by `space` or `enter` key
+   * dedicated `FocusConfig` can be set to autofocus first
+   * focusable element in popover container.
+   */
+
+  @HostListener('click', ['$event'])
+  @HostListener('keydown.enter', ['$event'])
+  handleOpen(event: MouseEvent | KeyboardEvent): void {
+    if (event.target === this.element.nativeElement) this.toggle(event);
+  }
+
+  /**
+   * Method performs open action for popover component.
+   */
+  open(event: MouseEvent | KeyboardEvent) {
+    if (!this.disablePopover) {
+      this.isOpen = true;
+      this.focusConfig = this.popoverService.getFocusConfig(
+        event,
+        this.appendToBody || false
+      );
+
+      this.renderPopover();
+
+      if (this.openPopover) this.openPopover.emit();
+    }
+  }
+
+  /**
+   * Method performs close action for popover component.
+   */
+  close() {
+    this.isOpen = false;
+    this.viewContainer.clear();
+
+    if (this.closePopover) this.closePopover.emit();
+  }
+
+  /**
+   * Method performs toggle action for popover component.
+   */
+  toggle(event: MouseEvent | KeyboardEvent) {
+    if (event && event.target === this.element.nativeElement && !this.isOpen)
+      this.open(event);
+    else if (this.isOpen) this.close();
+  }
+
+  /**
+   * Method subscribes for events emitted by popover component
+   * and based on event performs specific action.
+   */
+  handlePopoverEvents() {
+    this.popoverContainer.instance.eventSubject.subscribe(
+      (event: PopoverEvent) => {
+        if (event !== PopoverEvent.INSIDE_CLICK) this.close();
+        if (
+          event === PopoverEvent.ESCAPE_KEYDOWN ||
+          event === PopoverEvent.CLOSE_BUTTON_KEYDOWN
+        ) {
+          setTimeout(() =>
+            this.popoverService.setFocusOnElement(
+              this.element,
+              this.focusConfig,
+              this.appendToBody
+            )
+          );
+        }
+      }
+    );
+  }
+
+  /**
+   * Method creates instance and pass parameters to popover component.
+   */
+  renderPopover() {
+    const containerFactory = this.componentFactoryResolver.resolveComponentFactory(
+      PopoverComponent
+    );
+    this.popoverContainer = this.viewContainer.createComponent(
+      containerFactory
+    );
+    if (this.popoverContainer && this.popoverContainer.instance) {
+      this.popoverContainer.instance.content = this.cxPopover;
```

Comment:
Nitpick, what about 
`const componentInstance = this.popoverContainer.instance` 
and using this, to make it less boilerplated?